### DISCUSSION TITLE
Companion App and ffmpeg

### DISCUSSION THREAD

**Discussion by nollinvoyd** [aclap-dev/video-downloadhelper#153] [Help and Q&A]
### Your question

I ran across this nollinvoyd
Feb 17, 2024
Your question

I ran across this [[https://github.com/aclap-dev/vdhcoapp/actions/runs/7932289530](https://github.com/aclap-dev/video-downloadhelper/discussions/url)]([aclap-dev/vdhcoapp/actions/runs/793228953](https://github.com/aclap-dev/vdhcoapp/actions/runs/7932289530)

Is ffmpeg built into the Companion App?

**Comment by GrampaWildWilly** [comment]
I don't work for VDH.  I'm just another user.  I don't know the exact architecture of the CoApp.  But when you install the CoApp, it creates a directory that has several files in it, one of them being ffmpeg.exe.  So the program is freestanding.  I don't know how you would define "built into" when the program is a separately identifiable object.  However, the CoApp uses ffmpeg as its download engine.  That is a relatively recent development since last year.  Before that, for as long as there has been a CoApp (which is since Firefox 57), ffmpeg has been part of the CoApp package.  But its use was not for downloading.  It was used only for merging tracks into a complete video & conversions.  Those functions are still supported but like I said, ffmpeg now also does the downloads.

The CoApp is an open source project.  You can find the source code here on GitHub.  If you found the Actions section, you can surely find the source code without any help.  You might answer your own question that way.

Why are you asking?

**Reply by nollinvoyd** [reply]
I don't know why that link does not work. leading to page not found.

Thanks for the explanation.  That's what I figured.  I was curious because I have apps that depend on my ffmpeg stand alone version, and they must be directed to it's location.

"You might answer your own question that way" 

Anything's possible, but what is a novice like me going to do with source code? Not to mention, navigating Github for information is not user friendly for novices like me, i.e. trying finding the CoApp releases page there, which is why I visited there in the first place.

Fortunately I have a shortcut, because there is no "releases" tab.  There is a release page, but good luck digging around to find it.  Does that make sense? That's common for the projects there.  

Maybe it's intentional, to discourage the uninformed, like me, from participating and perhaps to seek help elsewhere, like here?

**Reply by GrampaWildWilly** [reply]
GitHub is kind of primarily for developers.  I probably couldn't read the code of the CoApp even if I looked at it.  Many useful links are in the Wiki section you can find by clicking its link at the top of this page.  In general, the Wiki sections of the repositories on GitHub are meant for documentation, so that's always a good place to look.

**Reply by GrampaWildWilly** [reply]
Also, Paul has posted in here that he has made some minor improvements to ffmpeg.  I have started using the CoApp's ffmpeg instead of my own standalone ffmpeg that I got from ffmpeg.org.  It seems to be working on everything I've tried it on so far.  I used to discourage people from doing that, but that was back in the days before the CoApp started using it for downloads.  Now, it's better suited to that purpose.

**Reply by nollinvoyd** [reply]
I understand that it's suited for developers, but why not at least make an effort to somewhat simplify things for those everyday users that must seek help here.  When you make things unnecessarily difficult, people will eventually look for alternatives elsewhere.  How. hard would it be to create a link/tab to releases?

**Reply by paulrouget** [reply]
> When you make things unnecessarily difficult
> How. hard would it be to create a link/tab to releases

Wait, I don't understand what you're talking about.
Everyday users don't go to coapp github, why would they do that?

You mean this? https://www.downloadhelper.net/install-coapp

That's the link we use in the addon for people to install or update the coapp.
That's the link that's all over the documentation, the website and github.

**Comment by nollinvoyd** [comment]
> GitHub is kind of primarily for developers. I probably couldn't read the code of the CoApp even if I looked at it. Many useful links are in the Wiki section you can find by clicking its link at the top of this page. In general, the Wiki sections of the repositories on GitHub are meant for documentation, so that's always a good place to look.

<details>
<summary>Details</summary>"I have started using the CoApp's ffmpeg instead of my own standalone ffmpeg".  

How do you go about that? You mean it's optional?



</details>

**Reply by paulrouget** [reply]
We provide a builtin ffmpeg.

You can use your own, but I do not recommend that as our version of ffmpeg comes with a bunch of fixes. If you're on linux, we have an easy way to do that: https://github.com/aclap-dev/video-downloadhelper/discussions/47

**Reply by GrampaWildWilly** [reply]
ffmpeg is an open source project completely separate from VDH.  If you want to learn more about it, visit ffmpeg.org.  I have been using ffmpeg as its own application to download certain things for a few years.  It's a tool that is documented in a way that only the people who wrote the code would consider useful.  What little I have learned about it began in a long thread on the old VDH forum on Google Groups that started during the pandemic when the Metropolitan Opera was posting a free opera every night.  People started talking about ffmpeg in that discussion & that's when I learned of its existence.  Since then, a lot of trial & error augmented by web searches has gotten me to where I can download most things I want to with ffmpeg.  The old forum still exists & you can search it for the Metropolitan Opera thread if you're interested.

Separately, I like to use VDH because it's so much simpler.  No need to tinker.  Push some buttons, give it a file name, & presto, the content is on my hard drive.

Now here's the overlap.  The VDH CoApp package includes a modified version of ffmpeg.  That's allowed.  The CoApp is open source.  ffmpeg is open source.  There's public licensing rules for that.  I don't know the details.  I just trust these guys to know all of that & make sure it's all legal & aboveboard.  That's what Paul is talking about above.  He's made some code corrections to ffmpeg.  Like I said above, ffmpeg didn't used to be the download engine inside the CoApp but now it is.  When I was first learning about ffmpeg, I tried to use the CoApp's ffmpeg to do some downloads & it just didn't work.  Back then, VDH/CoApp was not using ffmpeg for downloads so I suppose my results weren't surprising.  But things changed some time last summer when the CoApp started using ffmpeg for downloads.    It occurred to me that if the CoApp can use its captive ffmpeg to download content, that I should be able to now as well for my downloads that do not use VDH.  So that's what I've been doing.  Sometimes I use VDH & it does the heavy lifting to make ffmpeg work.  Sometimes I do my own heavy lifting & run the CoApp version of ffmpeg standalone for my own use.  I consider the CoApp's ffmpeg superior to the one I got from ffmpeg.org because of Paul's improvements.  So I use the CoApp ffmpeg for my non-VDH downloads.

Clearer?

**Reply by GrampaWildWilly** [reply]
Oh yeah.  One other thing.  Despite its developer bias, the Wiki link at the top of most discussions anywhere on GitHub, including the VDH part of GitHub, is meant for user documentation.  So if you have a question, you should be in the habit of first clicking that Wiki link & seeing what's there.  Maybe your question will be answered there.  If not, open a discussion just like you did here.

**Reply by nollinvoyd** [reply]
Got it. Thanks

**Comment by nollinvoyd** [comment]
> > Wait, I don't understand what you're talking about
> 

Sorry, that was really a generalized observation about Github, which is a great site, but there are projects here that invite everyday users, and could make things a little easier.

You guys have done a remarkable job.  Maybe once in a while we have something of value to offer, and you consider what we have to say.  Much appreciated :)

> > You mean this? [downloadhelper.net/install-coapp](https://www.downloadhelper.net/install-coapp)
>  That's the link we use in the addon for people to install or update the coapp. That's the link that's all >over the documentation, the website and github.
>

 I have a suggestion.  I didn't notice that existed until you pointed it out.  Maybe you could make it stand out by placing it among the tabs at the top, and/or changing the font color? Red?

**Reply by paulrouget** [reply]
Ah I wish we could :) but github doesn't allow that.

Thanks for the kind words though ❤️

**Comment by nollinvoyd** [comment]
> 
> Clearer?

So, you extract it from the installer?

**Reply by GrampaWildWilly** [reply]
The CoApp installer extracts it.  The installer creates a directory somewhere on one of your hard drives.  You can accept the default or place it anywhere you like.  I have not accepted the default.  Makes it easier to find.  In my case, that directory is C:\DownloadHelper CoApp.  That's actually just the default name, but I've promoted it to its own directory on C instead of it being buried somewhere inside the directory structure created by Windows install.  One of the files in that directory is already ffmpeg.exe.  I just execute that.  I don't copy it.  I don't clone it.  I just use it where it is.  Like this at a command prompt or in a .bat file:

`"C:\DownloadHelper CoApp\ffmpeg" parameters`

It's not all that hard to do.
